src/adapters/philips_hue/discovery.rs

```diff
@@ -0,0 +1,166 @@
+        let tx = self.adapter.tx.clone();
+        thread::spawn(move || {
+            let nupnp_enabled = controller.get_config().get_or_set_default(
+                "philips_hue", "nupnp_enabled", "true");
+            if nupnp_enabled == "true" {
+                let nupnp_url = controller.get_config().get_or_set_default(
+                    "philips_hue", "nupnp_url", "http://www.meethue.com/api/nupnp");
+                let nupnp_hubs = nupnp_query(&nupnp_url);
+                for nupnp in nupnp_hubs {
+                    let _ = tx.lock().unwrap().send(HueAction::AddHub(nupnp.id.to_owned(),
+                                nupnp.internalipaddress.to_owned()));
+                }
+            }
+        });
+    }
+
+    pub fn do_upnp_discovery(&self) {
+        let upnp = self.upnp_manager.lock().unwrap();
+        // Still wondering which one of these triggers a bridge response:
+        // upnp.search(Some("urn:schemas-upnp-org:device:basic:1".to_owned())).unwrap();
+        // upnp.search(Some("urn:schemas-upnp-org:device:Basic:1".to_owned())).unwrap();
+        // upnp.search(Some("urn:schemas-upnp-org:device:libhue:idl".to_owned())).unwrap();
+        // upnp.search(Some("upnp:rootdevice".to_owned())).unwrap();
+        // upnp.search(Some("uuid:2f402f80-da50-11e1-9b23-00178825681a".to_owned())).unwrap();
+        upnp.search(None).unwrap();
+    }
+}
+
+pub struct PhilipsHueUpnpListener<C> {
+    adapter: PhilipsHueAdapter<C>
+}
+
+impl<C: Controller> PhilipsHueUpnpListener<C> {
+    pub fn new(adapter: PhilipsHueAdapter<C>) -> Box<Self> {
+        Box::new(PhilipsHueUpnpListener {
+            adapter: adapter
+        })
+    }
+}
+
+impl<C: Controller> UpnpListener for PhilipsHueUpnpListener<C> {
+    // This will called each time that the device advertises itself using UPNP.
+    // A Philips Hue brisge posts an advertisement about a minute after search.
+    fn upnp_discover(&self, service: &UpnpService) -> bool {
+        macro_rules! try_get {
+            ($hash:expr, $key:expr) => (match $hash.get($key) {
+                Some(val) => val,
+                None => return false
+            })
+        }
+
+        let model_name = try_get!(service.description, "/root/device/modelName");
+
+        if !model_name.starts_with("Philips hue bridge") {
+            return false;
+        }
+
+        let serial = try_get!(service.description, "/root/device/serialNumber");
+        let url = try_get!(service.description, "/root/URLBase");
+        let model = try_get!(service.description, "/root/device/modelNumber");
+
+        debug!("UPnP announcement for Philips Hue bridge serial {} model {} at {}",
+            serial, model, url);
+
+        if !url.starts_with("http://") {
+            debug!("Unsupported URL scheme, expected HTTP.");
+            return false;
+        }
+
+        if !url.ends_with(":80/") {
+            debug!("Unsupported HTTP port, expected 80.");
```

Comment:
No, we don't need proper parsing for **UPnP** here. That check is probably even a bit over the top, because it's unlikely that Hue bridges will switch port any time soon. **nUPnP** doesn't even support listing the port at all.